### PR TITLE
List intersection group users

### PR DESCRIPTION
**Pull request by dvoet**
Ticket: <Link to Jira ticket>
<Put notes here to help reviewer understand this PR>
see PR comments I left for some guidance
---

**PR checklist**

- [ ] I've followed [the instructions](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#api-changes) if I've made any changes to the API, _especially_ if they're breaking changes
- [ ] I've updated the RC_XXX release ticket with any manual steps required to release this change
- [ ] I've updated the [FISMA documentation](https://github.com/broadinstitute/sam/blob/develop/CONTRIBUTING.md#fisma-documentation-changes) if I've made any security-related changes, including auth, encryption, or auditing

### CODE REVIEW

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala [hunk 1]
```diff
@@ -174,17 +174,7 @@ class PostgresDirectoryDAO(protected val dbRef: DbReference,
...
-
-      import SamTypeBinders._
-      withSQL {
-        select(groupTable.email)
-          .from(GroupTable as groupTable)
-          .where
-          .eq(groupTable.name, groupName)
-      }.map(rs => rs.get[WorkbenchEmail]("email")).single().apply()
-    }
+    batchLoadGroupEmail(Set(groupName)).map(_.toMap.get(groupName))
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala:177]: rewrote loadGroupEmail in terms of batchLoadGroupEmail

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala [hunk 2]
```diff
@@ -231,12 +221,23 @@ class PostgresDirectoryDAO(protected val dbRef: DbReference,
         case memberGroup: WorkbenchGroupIdentity =>
           val memberGroupPKQuery = workbenchGroupIdentityToGroupPK(memberGroup)
           samsql"insert into ${GroupMemberTable.table} (${groupMemberColumn.groupId}, ${groupMemberColumn.memberGroupId}) values ((${groupPKQuery}), (${memberGroupPKQuery}))"
-        case _ => throw new WorkbenchException(s"unexpected WorkbenchSubject $addMember")
+        case pet: PetServiceAccountId => throw new WorkbenchException(s"pet service accounts cannot be added to groups $pet")
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala:224]: more specific error message

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala [hunk 3]
```diff
@@ -231,12 +221,23 @@ class PostgresDirectoryDAO(protected val dbRef: DbReference,
...
       }
-      addMemberQuery.update().apply() > 0
+      val added = addMemberQuery.update().apply() > 0
+
+      if (added) {
+        updateGroupUpdatedDate(groupId)
+      }
+
+      added
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala:232]: update the updated date (same with remove member below)

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala [hunk 4]
```diff
@@ -302,21 +311,20 @@ class PostgresDirectoryDAO(protected val dbRef: DbReference,
...
-        UNION
-        SELECT ${gm.groupId}, ${gm.memberGroupId}, ${gm.memberUserId}
-        FROM sub_group sg, ${GroupMemberTable as gm}
-        WHERE ${gm.groupId} = sg.member_group_id
-      )
-      SELECT count(1)
-        FROM sub_group sg WHERE $memberClause"""
+      val query = samsql"""WITH RECURSIVE ${subGroupsQuery(subGroupMemberTable, topGroupQuery)}
+        SELECT count(1)
+        FROM ${subGroupMemberTable as sg} WHERE $memberClause"""
```
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala:311]: extracted some SQL for use in list intersection group members

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAOSpec.scala [hunk 5]
```diff
@@ -5,11 +5,11 @@ import org.broadinstitute.dsde.workbench.model.google.{GoogleProject, ServiceAcc
 import org.broadinstitute.dsde.workbench.sam.TestSupport
 import org.broadinstitute.dsde.workbench.sam.model.BasicWorkbenchGroup
 import org.postgresql.util.PSQLException
-import org.scalatest.{BeforeAndAfterEach, FlatSpec, Matchers}
+import org.scalatest.{BeforeAndAfterEach, FreeSpec, Matchers}
 
 import scala.concurrent.ExecutionContext.Implicits.global
 
-class PostgresDirectoryDAOSpec extends FlatSpec with Matchers with BeforeAndAfterEach {
+class PostgresDirectoryDAOSpec extends FreeSpec with Matchers with BeforeAndAfterEach {
```
Comment (dvoet) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAOSpec.scala:12]: The main thing to look at in this file is the conversion from FlatSpec to FreeSpec.

File: src/test/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAOSpec.scala [hunk 6]
```diff
@@ -21,259 +21,310 @@ class PostgresDirectoryDAOSpec extends FlatSpec with Matchers with BeforeAndAfte
...
+      "return true when policy is in policy" is pending
+      "return false when policy is not in policy" is pending
+      "return true when policy is in group" is pending
+      "return false when policy is not in group" is pending
+      "return true when group is in policy" is pending
+      "return false when group is not in policy" is pending
+    }
 
-    dao.isGroupMember(parentGroup.id, subGroup1.id).unsafeRunSync() should be (false)
+    "listIntersectionGroupUsers" - {
```
Comment (dvoet) [inline src/test/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAOSpec.scala:304]: this is the only new test in this PR

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala (hunk 1 above)
Comment (dvoet) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala]: rewrote loadGroupEmail in terms of batchLoadGroupEmail

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala [hunk 7]
```diff
@@ -281,12 +282,20 @@ class PostgresDirectoryDAO(protected val dbRef: DbReference,
...
     }
   }
 
   override def isGroupMember(groupId: WorkbenchGroupIdentity, member: WorkbenchSubject): IO[Boolean] = {
     val gm = GroupMemberTable.syntax("gm")
+    val subGroupMemberTable = SubGroupMemberTable("sub_group")
+    val sg = subGroupMemberTable.syntax("sg")
 
     val memberClause: SQLSyntax = member match {
       case subGroupId: WorkbenchGroupIdentity => samsqls"sg.member_group_id = (${workbenchGroupIdentityToGroupPK(subGroupId)})"
```
Comment (gpolumbo-broad) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala]: I think you can now replace `samsqls"sg.member_group_id..."` with `samsqls"${sg.memberGroupId}..."`

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala [hunk 8]
```diff
@@ -302,21 +311,20 @@ class PostgresDirectoryDAO(protected val dbRef: DbReference,
...
-        $topGroupQuery
-        UNION
-        SELECT ${gm.groupId}, ${gm.memberGroupId}, ${gm.memberUserId}
-        FROM sub_group sg, ${GroupMemberTable as gm}
-        WHERE ${gm.groupId} = sg.member_group_id
-      )
-      SELECT count(1)
-        FROM sub_group sg WHERE $memberClause"""
+      val query = samsql"""WITH RECURSIVE ${subGroupsQuery(subGroupMemberTable, topGroupQuery)}
+        SELECT count(1)
```
Comment (gpolumbo-broad) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala]: May not want to use `count(1)`:  https://stackoverflow.com/questions/2710621/count-vs-count1-vs-countpk-which-is-better

May want to consider using `EXISTS` too:  https://www.postgresql.org/docs/8.1/functions-subquery.html

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala [hunk 9]
```diff
@@ -381,7 +389,57 @@ class PostgresDirectoryDAO(protected val dbRef: DbReference,
...
+
+    // the toSeq below is important to fix a predictable order
+    val subGroupQueries = groupIds.toSeq.zipWithIndex.map { case (groupId, index) =>
+      // need each subgroup table to be named uniquely
+      // this is careful not to use a user defined string (e.g. the group's name) to avoid sql injection attacks
+      val subGroupTable = SubGroupMemberTable("sub_group_" + index)
+      val topGroupQuery =
+        samsqls"""select ${gm.groupId}, ${gm.memberGroupId}, ${gm.memberUserId}
+                 from ${GroupMemberTable as gm}
+                 where ${gm.groupId} = (${workbenchGroupIdentityToGroupPK(groupId)})"""
```
Comment (gpolumbo-broad) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala]: Same `topGroupQuery` as used on line 306 above: https://github.com/broadinstitute/sam/pull/318/files#diff-1bab66f5a3fc16258167e043394e5fcdR306 - consider extracting

File: src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala [hunk 10]
```diff
@@ -381,7 +389,78 @@ class PostgresDirectoryDAO(protected val dbRef: DbReference,
...
+    * each `SELECT` that creates each individual CTE.  In this implementation, we use `WITH RECURSIVE` to create N CTEs
+    * where N is the cardinality of the groupIds parameter.  Each of these CTEs is the flattened group membership for
+    * one of the WorkbenchGroupIdentities.
+    *
+    * The query maintains a table alias for each of the CTEs.  The final part of the query uses this list of aliases to
+    * join each of them based on the memberUserId column to give us the final result of only those memberUserIds that
+    * showed up as an entry in every CTE.
+    * @param groupIds
+    * @return Set of WorkbenchUserIds that are members of each group specified by groupIds
+    */
```
Comment (gpolumbo-broad) [inline src/main/scala/org/broadinstitute/dsde/workbench/sam/directory/PostgresDirectoryDAO.scala:407]: @dvoet - does this explanation sound correct to you?

**Review by marctalbott — APPROVED** [review]

**Review by gpolumbo-broad — APPROVED** [review]

**Pull request merged**